bin/faucet/src/client.rs

```diff
@@ -197,6 +200,9 @@ impl FaucetDataStore {
     }
 }
 
+unsafe impl Send for FaucetDataStore {}
+unsafe impl Sync for FaucetDataStore {}
```

Comment:
I've implemented these two traits for the `FaucetDataStore` as it was done for other structures which were implementing the `DataStore` trait (for example, [here](https://github.com/0xPolygonMiden/miden-base/blob/ea1de499461aa27a951309bd7a725dc7ddc631dc/miden-tx/src/testing/tx_context/mod.rs#L124)), but I'm not sure whether it is needed here. With these implementations we remove one of two warnings about using a structure which is not implementing these traits in `Arc`, but we allow this warning anyway.

Reply:
These have always bothered me but I don't know the history/context behind why we need them.

Do you have an understanding of why the `Send` and `Sync` aren't automatically implemented for these types? If so we should definitely add `/// SAFETY` comments explaining this.

Ideally these shouldn't be required at all. And I'm confused why implementing another trait causes these auto-traits to disappear :D

Reply:
While this is working here due to serial nature of our faucet web app, I think, the right solution would be to update storage itself in order to implement Send/Sync automatically: https://github.com/0xPolygonMiden/miden-node/pull/517/files#diff-1d5b7b3cf77eb4b236ed8fec8e25c2ea541e3591c719f9b45fae328c4e1ada26R13-R18

If `Sync` is also required for `FaucetDataStore`, we will probably need to change `RwLock` to `Mutex`.

Reply:
> While this will work here due to a serial nature of our faucet web app, I do think, the proper solution would be to update storage itself in order to implement `Send` automatically: https://github.com/0xPolygonMiden/miden-node/pull/517/files#diff-1d5b7b3cf77eb4b236ed8fec8e25c2ea541e3591c719f9b45fae328c4e1ada26R13-R18
> 
> If `Sync` is also required for `FaucetDataStore`, we will probably need to change `RwLock` to `Mutex`.

This sounds good to me; and I think we may as well make it `Sync`. I don't think it will impact the faucet throughput.

Reply:
What prevents storage from implementing `Send` automatically? And `RwLock` should be `Sync` so I'm not following 🥲

Reply:
Maybe there are borrowed references which require it to implement `Sync` in order to be `Send`, but `RefCell` doesn't implement `Sync`. `RwLock` implements `Sync` only it its content implement `Sync` as well, so it's likely that `Arc<RwLock<...>>` should be enough here.

Reply:
You can see differences in `Send`/`Sync` implementation for mentioned structs here:

RefCell: https://doc.rust-lang.org/src/core/cell.rs.html#1280-1284
RwLock: https://doc.rust-lang.org/src/std/sync/rwlock.rs.html#89-92
Mutex: https://doc.rust-lang.org/src/std/sync/mutex.rs.html#186-189

Reply:
Using `Mutex` for the `faucet_account` should fix this.